### PR TITLE
Validation rule checking if access port is associated with just single network

### LINKED ISSUE #695: Create a validation rule that would prevent attaching multiple networks to an access port
### Which role would this new feature be associated?

cisco.nac_dc_vxlan.validate

### New feature description:

```shell
Access port is an interface that can have only single network/vlan attached.
I would be good to have a rule built-in that would guard that and ensure that only a single vlan is attached to an access port.
```

### Would this be a new feature requiring a new role?

```shell
no
```

### Which section of the data model is the new feature related?

vxlan.overlay

### Would this be a new feature requiring a new addition to the data model?

```shell
no
```

### New or Affected Documentation for New Feature

```shell

```

### Potential Ansible Task Config

```yaml
vxlan:
  topology:
    switches:
      - name: LEAF_11
        interfaces:
          - name: Ethernet1/15
            mode: access
            access_vlan: 101
            speed: 10Gb
            mtu: default
            enabled: true
  overlay:
    networks:
      - name: PROD_300
        net_id: 30000
        vrf_name: PROD_VRF
        vlan_id: 300
        vlan_name: PROD_300
        gw_ip_address: 10.30.0.1/24
        is_l2_only: false
        arp_suppress: false
        int_desc: PROD_300
        mtu_l3intf: 9100
        network_attach_group: prod_300
      - name: DEV_2002
        net_id: 30002
        vrf_name: DEV_VRF
        vlan_id: 2002
        vlan_name: DEV_2002
        gw_ip_address: 10.200.2.1/24
        is_l2_only: false
        arp_suppress: false
        dhcp_servers:
          - ip_address: 192.168.200.200
            vrf: management
        int_desc: DEV_2002
        mtu_l3intf: 9216
        network_attach_group: dev_2002
    network_attach_groups:
      - name: prod_300
        switches:
          - hostname: LEAF_11
            ports:
              - Ethernet1/15
      - name: dev_2002
        switches:
          - hostname: LEAF_11
            ports:
              - Ethernet1/15
```

### Ansible Version

```shell

```

### Ansible Collection Versions

```shell

```

### Cisco Nexus Dashboard Version

```shell

```

### Cisco NX-OS Version

```shell

```

### PR DESCRIPTION
**Pull request by mkilar123**
## Related Issue(s)
Fixes #695 

## Related Collection Role

* [x] cisco.nac_dc_vxlan.validate
* [ ] cisco.nac_dc_vxlan.dtc.create
* [ ] cisco.nac_dc_vxlan.dtc.deploy
* [ ] cisco.nac_dc_vxlan.dtc.remove
* [ ] other

## Related Data Model Element

* [ ] vxlan.fabric
* [ ] vxlan.global
* [ ] vxlan.topology
* [ ] vxlan.underlay
* [ ] vxlan.overlay
* [ ] vxlan.overlay_extensions
* [ ] vxlan.policy
* [ ] vxlan.multisite
* [ ] defaults.vxlan
* [ ] other

## Proposed Changes
Adding a validation rule which checks if an access interface is mapped to maximum one network:
Checked conditions:
- If the access interface has "access_vlan" defined - it cannot be associated with any network_attach_group
- If the access interface has no "access_vlan" defined - it can be associated with maximum one network_attach_group
- If network_attach_group contains an access port - it can be associated with maximum one network
The checks also apply to interfaces on TORs

## Test Notes
Tests run:
- access interface without "access_vlan" defined referenced max once in network_attach_groups
- access interface with"access_vlan" defined not referenced in any network_attach_groups
- TOR access interface without "access_vlan" defined referenced max once in network_attach_groups
- TOR access interface with"access_vlan" defined not referenced in any network_attach_groups

- access interface with"access_vlan" defined referenced only once in network_attach_groups
- TOR access interface with"access_vlan" defined referenced only once in network_attach_groups

- access interface without "access_vlan" defined referenced multiple times in network_attach_groups
- TOR access interface without "access_vlan" defined referenced multiple times in network_attach_groups

- network_attach_group containing access port referenced by one network
-  network_attach_group containing access port referenced by multiple networks
- - network_attach_group without access port referenced by multiple networks

## Checklist

* [x] Latest commit is rebased from develop with merge conflicts resolved
* [ ] New or updates to documentation has been made accordingly
* [x] Assigned the proper reviewers

### CODE REVIEW

**Review by juburnet — COMMENTED** [review]
# Comprehensive Review: PR #690 - Validation Rule for Access Interface VLAN Assignments

## 🎯 Overview
This PR introduces a new HIGH severity validation rule (ID: 405) to enforce proper access interface to network attach group mappings. The implementation addresses a critical configuration gap for both switch and TOR topologies.

---

## ✅ Strengths

1. **Clear business logic**: The rule enforces three distinct constraints that prevent VLAN configuration conflicts
2. **Comprehensive coverage**: Handles both switch (2-tuple) and TOR (3-tuple) interface references
3. **Descriptive error messages**: Each validation failure provides actionable context (hostname, interface, networks)
4. **Defensive programming**: Uses utility methods `data_model_key_check()` and `safeget()` for safe dictionary access
5. **Well-documented**: Docstrings explain the purpose of each helper method
6. **CI passing**: All 23 checks pass, indicating no regressions

---

## 🔴 Critical Issues (Must Fix Before Merge)

### 1. Missing None Check in Interface Info Access (Lines 237, 277)
**Location**: `validate_access_interface_references()`

**Issue**: After retrieving `interface_info` from `access_interfaces_map`, the code immediately accesses `interface_info['has_access_vlan']` without validating the dictionary structure.

**Risk**: If `interface_info` is an empty dict or malformed, this will raise a `KeyError`.

**Recommendation**:
```python
interface_info = access_interfaces_map[switch_key]
has_access_vlan = interface_info.get('has_access_vlan', False)
if has_access_vlan is None:
    # Log warning or skip
    continue
```

---

## 🟡 Should Fix (Not Blocking)

### 2. Code Duplication: Switch vs TOR Validation (Lines 233-259 vs 261-301)
**Issue**: Nearly identical validation logic repeated for switches and TORs with only minor string formatting differences.

**Impact**: Maintenance burden - bug fixes must be applied twice.

**Recommendation**: Extract common validation logic into a shared helper method:
```python
@classmethod
def _validate_interface_vlan_reference(cls, hostname, interface_name, has_access_vlan, 
                                        reference_count, group_names, device_type='switch', parent_leaf=None):
    """
    Shared validation logic for both switches and TORs
    """
    # Common validation with parameterized error messages
```

### 3. Inefficient Iteration Pattern (Lines 138-205)
**Issue**: The code iterates through all interfaces and then filters by `mode == 'access'`, rather than pre-filtering.

**Impact**: Minor performance overhead when processing large switch configs.

**Recommendation**: If the data model allows, consider pre-filtering access interfaces before iteration.

---

## 🟢 Consider for Improvement

### 4. Confusing Variable Naming (Line 263)
**Location**: TOR validation block

**Issue**: Variable `group_names` actually contains network names, not group names:
```python
for ref_key, group_names in interface_references.items():
    # ...
    network_list = ', '.join(group_names)  # These are network names, not group names
```

**Recommendation**: Rename to `network_names` for clarity.

### 5. Utility Method Redundancy (Lines 343-373)
**Issue**: `data_model_key_check()` and `safeget()` may duplicate functionality from other validation rules.

**Recommendation**: Check if similar utilities exist in other rules (e.g., 401, 402, 403, 404). If so, consider extracting to a shared `validation_utils.py` module.

---

## 💭 Questions / Needs Discussion

### 6. Validation Coverage: Trunk Ports
**Question**: This rule only validates `access` mode interfaces. Should there be validation for trunk ports that might also reference network_attach_groups?

**Context**: Line 155 explicitly filters for `mode == 'access'`. Is this intentional scope, or should trunk ports be covered in a separate rule?

### 7. Error Reporting: Aggregation vs Individual
**Observation**: The rule appends individual error messages to `cls.results`. For large deployments with many violations, this could produce hundreds of error lines.

**Question**: Should there be a configurable limit or summary aggregation (e.g., "Found 47 access interface violations across 12 switches. See details below.")?

---

## 🧪 Testing Recommendations

### Suggested Test Cases (If Not Already Covered):
1. **Edge case**: Access interface with `access_vlan` referenced in multiple network attach groups
2. **Edge case**: TOR interface referenced from multiple leaf switches
3. **Edge case**: Network attach group referenced by 3+ networks (not just 2)
4. **Edge case**: Empty `access_interfaces` list on a switch
5. **Edge case**: Malformed interface definition missing expected keys
6. **Regression**: Ensure rule doesn't fire false positives on trunk interfaces

---

## 📊 Code Quality Metrics

| Metric | Value | Status |
|--------|-------|--------|
| Lines Added | 367 | ✅ |
| Lines Deleted | 0 | ✅ (New feature) |
| Cyclomatic Complexity | Medium | ⚠️ (Could be reduced with refactoring) |
| Documentation | Good | ✅ |
| Error Messages | Excellent | ✅ |
| Test Coverage | Unknown | ❓ (No test files in PR) |

---

## 🎬 Recommendation

**Status**: ⚠️ **REQUEST CHANGES**

**Reason**: Critical issue #1 (missing None check) could cause runtime errors in edge cases.

**Action Items**:
1. 🔴 Add defensive None/KeyError handling for `interface_info['has_access_vlan']` access
2. 🟡 Consider refactoring duplicated validation logic
3. 💭 Clarify scope: Should trunk ports be validated in a future rule?

Once issue #1 is addressed, this will be **APPROVED** - the core logic is sound and the feature is valuable.

---

## 🔗 Related Context
- Related Issue: #695
- Rule ID: 405
- Severity: HIGH
- Applies to: `common_vxlan` validation

File: roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+            # Check if this interface is referenced in network attach groups as a direct port
+            referenced_networks = interface_references.get(key, [])
+            reference_count = len(referenced_networks)
+
+            if has_access_vlan:
+                # Interface has access_vlan defined - it cannot be referenced in network attach groups
+                if reference_count > 0:
+                    network_list = ', '.join(referenced_networks)
+                    cls.results.append(
```
Comment (juburnet) [inline roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py]: 🔴 **Critical: Missing None/KeyError Check**

Directly accessing `interface_info['has_access_vlan']` without validation could raise a `KeyError` if the dictionary structure is malformed or missing expected keys.

**Recommendation**: Use defensive access:
```suggestion
has_access_vlan = interface_info.get('has_access_vlan', False)
if has_access_vlan is None:
    continue
```

This ensures the code gracefully handles unexpected data structures instead of crashing at runtime.

File: roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                continue
+
+            interface_info = access_interfaces_map[tor_key]
+            has_access_vlan = interface_info['has_access_vlan']
+
+            reference_count = len(group_names)
+
+            if has_access_vlan:
+                # Interface has access_vlan defined - it cannot be referenced in network attach groups
+                if reference_count > 0:
```
Comment (juburnet) [inline roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py]: 🔴 **Critical: Same KeyError Risk (Duplicate Issue)**

Same issue as line 237 - direct dictionary access without validation.

**Recommendation**: Apply the same fix:
```suggestion
has_access_vlan = interface_info.get('has_access_vlan', False)
if has_access_vlan is None:
    continue
```

File: roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+            has_access_vlan = interface_info.get('has_access_vlan', False)
+
+            # Skip if essential fields are missing
+            if not hostname or not interface_name:
+                cls.results.append(
+                    f"Malformed interface data in access_interfaces_map: missing hostname or interface_name. "
+                    f"Key: {key}, Data: {interface_info}. Skipping validation for this interface."
+                )
+                continue
+
```
Comment (juburnet) [inline roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py]: ✅ **Critical Issue Resolved**

Excellent work adding this defensive validation! This addresses the critical issue from the previous review (2026-02-12).

The code now properly handles malformed interface data by:
1. Checking for missing `hostname` or `interface_name`
2. Logging a descriptive error message
3. Skipping validation for that interface with `continue`

This prevents the potential KeyError that was the main blocking concern in the previous review.

**Impact**: This change makes the validation rule robust against data model inconsistencies and prevents runtime exceptions.

File: roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+            referenced_networks = interface_references.get(key, [])
+            reference_count = len(referenced_networks)
+
+            if has_access_vlan:
+                # Interface has access_vlan defined - it cannot be referenced in network attach groups
+                if reference_count > 0:
+                    network_list = ', '.join(referenced_networks)
+                    cls.results.append(
+                        f"Access interface '{interface_name}' on switch '{hostname}' has 'access_vlan' defined "
+                        f"and cannot be referenced in network_attach_groups. "
```
Comment (juburnet) [inline roles/validate/files/rules/common_vxlan/405_overlay_access_interfaces.py]: 🟡 **Code Duplication Opportunity**

The validation logic here (lines 247-267) is nearly identical to the TOR validation logic (lines 289-309). The only difference is error message formatting.

**Impact**: Maintenance burden - any bug fixes or logic changes must be applied in two places.

**Suggestion**: Consider extracting common validation into a helper method. This refactoring can be done in a follow-up PR if you prefer to keep this PR focused on the core functionality.

Example approach:
```python
@classmethod
def _validate_interface_vlan_constraint(cls, hostname, interface_name, has_access_vlan, 
                                         reference_count, referenced_networks, 
                                         device_type='switch', parent_leaf=None):
    # Shared validation logic with parameterized error messages
    ...
```
